services/acm/src/main/java/software/amazon/awssdk/services/acm/auth/scheme/internal/AcmAuthSchemeInterceptor.java

```diff
@@ -0,0 +1,132 @@
+ * and limitations under the License.
+ */
+
+package software.amazon.awssdk.services.acm.auth.scheme.internal;
+
+import java.util.List;
+import java.util.Map;
+import software.amazon.awssdk.annotations.Generated;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.awscore.AwsExecutionAttribute;
+import software.amazon.awssdk.core.SelectedAuthScheme;
+import software.amazon.awssdk.core.interceptor.Context;
+import software.amazon.awssdk.core.interceptor.ExecutionAttributes;
+import software.amazon.awssdk.core.interceptor.ExecutionInterceptor;
+import software.amazon.awssdk.core.interceptor.SdkExecutionAttribute;
+import software.amazon.awssdk.http.auth.spi.AuthScheme;
+import software.amazon.awssdk.http.auth.spi.AuthSchemeOption;
+import software.amazon.awssdk.http.auth.spi.IdentityProviderConfiguration;
+import software.amazon.awssdk.identity.spi.Identity;
+import software.amazon.awssdk.identity.spi.IdentityProvider;
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.services.acm.auth.scheme.AcmAuthSchemeParams;
+import software.amazon.awssdk.services.acm.auth.scheme.AcmAuthSchemeProvider;
+import software.amazon.awssdk.utils.Logger;
+
+@Generated("software.amazon.awssdk:codegen")
+@SdkInternalApi
+public final class AcmAuthSchemeInterceptor implements ExecutionInterceptor {
+
+    private static final Logger log = Logger.loggerFor(AcmAuthSchemeInterceptor.class);
+
+    @Override
+    public void beforeExecution(Context.BeforeExecution context, ExecutionAttributes executionAttributes) {
+        List<AuthSchemeOption> authOptions = resolveAuthOptions(executionAttributes);
+        SelectedAuthScheme<?> selectedAuthScheme = selectAuthScheme(authOptions, executionAttributes);
+        executionAttributes.putAttribute(SdkExecutionAttribute.SELECTED_AUTH_SCHEME, selectedAuthScheme);
+    }
+
+    /**
+     * Invoke the auth scheme resolver to determine which auth options we should consider for this request.
+     */
+    private List<AuthSchemeOption> resolveAuthOptions(ExecutionAttributes executionAttributes) {
+        // Prepare the inputs for the auth scheme resolver. We always include the
+        // operationName, and we include the region if the service is modeled with
+        // @sigv4.
+        String operation = executionAttributes.getAttribute(SdkExecutionAttribute.OPERATION_NAME);
+        // TODO: Do this only if sigv4
+        Region region = executionAttributes.getAttribute(AwsExecutionAttribute.AWS_REGION);
+
+        // Use the operation-level auth scheme resolver if the customer
+        // specified an override, otherwise fall back to the one on the client.
+        // TODO: The fallback should happen while setting in execution attributes
+        AcmAuthSchemeProvider authSchemeProvider =
+            (AcmAuthSchemeProvider) executionAttributes.getAttribute(SdkExecutionAttribute.AUTH_SCHEME_RESOLVER);
```

Comment:
Is the `AUTH_SCHEME_RESOLVER` needed outside of this module?

If not, should we have a codegen'd `AcmExecutionAttribute.AUTH_SCHEME_RESOLVER`? That would allow us to make it `AcmAuthSchemeProvider` so that it's type-safe through the whole execution, and it would also let us make it an `@SdkInternalApi` instead of an `@SdkProtectedApi`.

Reply:
Like the idea, but AUTH_SCHEME_RESOLVER execution attribute needs to be set by service agnostic AwsExecutionContextBuilder, for the AuthSchemeInterceptor to read, which is needing the cast here. But I'm changing the exeuctionAttribute to be SdkInternalExecutionAttribute and is only set by reading the clientConfig, which will be set via a setter in the generated service specific client builder with {Service}AuthSchemeProvider, so the cast is safe?

This is the same pattern used for EndpointProvider.

Reply:
Okay, hopefully we can clean it up with the smithy migration when things get pushed down into the clients.